File: lib/ts/src/internal/generation/GenerateRandomSelect.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { GenerateContext } from './GenerateContext.js';

export function generateRandomSelect(possibleSelects: Record<string, any>, ctx: GenerateContext): object {
    const possibleSelect = possibleSelects[ctx.fnScope];
    return subSelect(possibleSelect, ctx);
}

function subSelect(possibleSelectSection: any, ctx: GenerateContext): any {
    if (Array.isArray(possibleSelectSection)) {
        const selectedFieldNames = [];

        for (const fieldName of possibleSelectSection) {
            if (ctx.randomGenerator.nextBoolean()) {
                selectedFieldNames.push(fieldName);
            }
        }

        return selectedFieldNames.sort();
    } else if (typeof possibleSelectSection === 'object' && !Array.isArray(possibleSelectSection)) {
        const selectedSection: Record<string, any> = {};

        const keys = Object.keys(possibleSelectSection).sort();

        for (const key of keys) {
            const value = possibleSelectSection[key];
            if (ctx.randomGenerator.nextBoolean()) {
                const result = subSelect(value, ctx);
                if (typeof result === 'object' && !Array.isArray(result) && Object.keys(result).length === 0) {
                    continue;
                }
                selectedSection[key] = result;
            }
        }

        return selectedSection;
    } else {
        throw new Error('Invalid possibleSelectSection');
    }
}

File: lib/go/telepact/internal/types/GenerateRandomSelect.go
//|
//|  Copyright The Telepact Authors
//|
//|  Licensed under the Apache License, Version 2.0 (the "License");
//|  you may not use this file except in compliance with the License.
//|  You may obtain a copy of the License at
//|
//|  https://www.apache.org/licenses/LICENSE-2.0
//|
//|  Unless required by applicable law or agreed to in writing, software
//|  distributed under the License is distributed on an "AS IS" BASIS,
//|  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//|  See the License for the specific language governing permissions and
//|  limitations under the License.
//|

package types

import (
	"fmt"
	"sort"
)

// GenerateRandomSelect builds a pseudo-random select projection using the provided context.
func GenerateRandomSelect(possibleSelects map[string]any, ctx *GenerateContext) any {
	if ctx == nil {
		return nil
	}

	possibleSelect := possibleSelects[ctx.FnScope]
	return subSelect(possibleSelect, ctx)
}

func subSelect(possibleSelectSection any, ctx *GenerateContext) any {
	switch typed := possibleSelectSection.(type) {
	case []any:
		selectedFieldNames := make([]string, 0)
		for _, entry := range typed {
			fieldName, ok := entry.(string)
			if !ok {
				continue
			}

			if ctx.RandomGenerator == nil || ctx.RandomGenerator.NextBoolean() {
				selectedFieldNames = append(selectedFieldNames, fieldName)
			}
		}
		sort.Strings(selectedFieldNames)
		return selectedFieldNames
	case []string:
		selectedFieldNames := make([]string, 0, len(typed))
		for _, fieldName := range typed {
			if ctx.RandomGenerator == nil || ctx.RandomGenerator.NextBoolean() {
				selectedFieldNames = append(selectedFieldNames, fieldName)
			}
		}
		sort.Strings(selectedFieldNames)
		return selectedFieldNames
	case map[string]any:
		selectedSection := make(map[string]any)
		keys := make([]string, 0, len(typed))
		for key := range typed {
			keys = append(keys, key)
		}
		sort.Strings(keys)

		for _, key := range keys {
			value := typed[key]
			if ctx.RandomGenerator != nil && !ctx.RandomGenerator.NextBoolean() {
				continue
			}

			result := subSelect(value, ctx)
			if nestedMap, ok := result.(map[string]any); ok && len(nestedMap) == 0 {
				continue
			}
			selectedSection[key] = result
		}

		return selectedSection
	case map[string][]string:
		converted := make(map[string]any, len(typed))
		for key, value := range typed {
			converted[key] = value
		}
		return subSelect(converted, ctx)
	case map[any]any:
		converted := make(map[string]any, len(typed))
		for key, value := range typed {
			converted[fmt.Sprintf("%v", key)] = value
		}
		return subSelect(converted, ctx)
	default:
		panic(fmt.Errorf("invalid possible_select_section: %T", possibleSelectSection))
	}
}
